plugins/uefi-capsule/fu-uefi-bootmgr.c

```diff
@@ -249,6 +249,151 @@ fu_uefi_setup_bootnext_with_loadopt(FuEfiLoadOption *loadopt, GError **error)
 	return TRUE;
 }
 
+static gboolean
+fu_uefi_bootmgr_get_generation(FuCsvEntry *entry, guint64 *generation, GError **error)
+{
+	const gchar *generation_str =
+	    fu_csv_entry_get_value_by_column_id(entry, "component_generation");
+
+	if (generation_str == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_FILE,
+			    "missing component generation");
+
+		return FALSE;
+	}
+
+	if (!fu_strtoull(generation_str, generation, 0, G_MAXUINT, error)) {
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_uefi_bootmgr_shim_is_safe(const gchar *source_shim, GError **error)
+{
+	gsize idx = 0;
+	g_autoptr(GBytes) current_sbatlevel_bytes = NULL;
+	g_autoptr(FuFirmware) shim = NULL;
+	g_autoptr(FuFirmware) sbatlevel_section = NULL;
+	g_autoptr(FuFirmware) previous_sbatlevel = NULL;
+	g_autoptr(FuFirmware) current_sbatlevel = NULL;
+	g_autoptr(GError) error_local = NULL;
+	g_autoptr(GBytes) blob = NULL;
+	g_autoptr(GBytes) sbatlevel_bytes = NULL;
+	g_autoptr(GPtrArray) shim_entries = NULL;
+
+	shim = fu_pefile_firmware_new();
+	blob = fu_bytes_get_contents(source_shim, error);
+	if (!blob)
+		return FALSE;
+	if (!fu_firmware_parse(shim, blob, FWUPD_INSTALL_FLAG_NONE, error))
+		return FALSE;
+
+	sbatlevel_section = fu_firmware_get_image_by_id(shim, ".sbatlevel", &error_local);
+	if (sbatlevel_section == NULL) {
+		g_debug("no sbatlevel section was found");
+		/* if there is no .sbatlevel section, then it will not update, it should be safe */
+		if (g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_NOT_FOUND))
+			return TRUE;
+		g_propagate_error(error, g_steal_pointer(&error_local));
+	}
+
+	previous_sbatlevel = fu_firmware_get_image_by_id(sbatlevel_section, "previous", error);
+	if (previous_sbatlevel == NULL)
+		return FALSE;
+
+	shim_entries = fu_firmware_get_images(previous_sbatlevel);
+
+	current_sbatlevel_bytes =
+	    fu_efivar_get_data_bytes(FU_EFIVAR_GUID_SHIM, "SbatLevelRT", NULL, error);
+	if (current_sbatlevel_bytes == NULL) {
+		/* not safe if variable is not set but new shim would set it */
```

Comment:
can you move the comment to before `current_sbatlevel_bytes =` and remove the braces pls